gpytorch/models/exact_prediction_strategies.py

```diff
@@ -73,6 +76,78 @@ def _exact_predictive_covar_inv_quad_form_root(self, precomputed_cache, test_tra
         # where S S^T = (K_XX + sigma^2 I)^-1
         return test_train_covar.matmul(precomputed_cache)
 
+    def get_fantasy_strategy(self, inputs, targets, full_inputs, full_targets, full_output):
+        full_mean, full_covar = full_output.mean, full_output.lazy_covariance_matrix
+
+        batch_shape = full_inputs[0].shape[:-2]
+
+        full_mean = full_mean.view(*batch_shape, -1)
+        num_train = self.num_train
+
+        # Evaluate fant x train and fant x fant covariance matrices, leave train x train unevaluated.
+        fant_fant_covar = full_covar[..., num_train:, num_train:]
+        mvn = self.likelihood(MultivariateNormal(torch.zeros(1), fant_fant_covar), inputs)
+        fant_fant_covar = mvn.covariance_matrix
+
+        fant_train_covar = full_covar[..., num_train:, :num_train].evaluate()
+        fant_mean = full_mean[..., num_train:]
+
+        self.fantasy_inputs = inputs
+        self.fantasy_targets = targets
+
+        # Get cached K inverse decomp. (or compute if we somehow don't already have the covariance cache)
+        K_inverse = self.lik_train_train_covar.root_inv_decomposition()
+        fant_solve = K_inverse.matmul(fant_train_covar.transpose(-2, -1))
+
+        schur_complement = fant_fant_covar - fant_train_covar.matmul(fant_solve)
+        small_system_rhs = targets - fant_mean - fant_train_covar.matmul(self.mean_cache)
+
+        # Schur complement of a spd matrix is guaranteed to be positive definite
+        if small_system_rhs.requires_grad or schur_complement.requires_grad:
+            # TODO: Delete this part of the if statement when PyTorch implements potrs derivative.
+            fant_cache_lower = torch.gesv(small_system_rhs.unsqueeze(-1), schur_complement)[0].squeeze(-1)
+        else:
+            fant_cache_lower = torch.potrs(small_system_rhs, torch.cholesky(schur_complement, upper=True)).squeeze(-1)
+
+        fant_cache_upper = self.mean_cache - fant_solve.matmul(fant_cache_lower)
+
+        fant_mean_cache = torch.cat((fant_cache_upper, fant_cache_lower), dim=-1)
+
+        # [K U; U S] = [L 0; lower_left schur_root]
+        L_inverse = self.covar_cache
+        L = self.lik_train_train_covar.root_decomposition().root.evaluate()
+        lower_left = fant_train_covar.matmul(L_inverse)
+        schur_root = torch.cholesky(fant_fant_covar - lower_left.matmul(lower_left.transpose(-2, -1)))
+        upper_right = torch.zeros(L.size(-2), schur_root.size(-1))
+
+        # Form pseudo-inverse of [L 0; lower_left schur_root]
```

Comment:
instead of concatenating the tensors, how about allocating `new_root` as a zero tensor of appropriate size and then filing the non-zero elements via indexed assignment? Not sure if that makes much of a difference.

Reply:
I think this will mess with autograd as we'll be operating on the tensor in-place. I'll try it though.